docs/0002-N-tuple-tree.md

```diff
@@ -0,0 +1,180 @@
+
+## Overview
+
+This extension provides a general mechanism for describing a set of algorithms for mapping numerical identifers onto a tree structure for constructing an OCFL Storage Hierarchy in a way that maintains performance. The aim is to balance the problem of having too many files or subdirectories in any one directory with having an overly deeply nested set of subdirectories. As both identifier formats and specific filesystem instances may vary there is not a single optimal approach.       
+
+## Parameters
+
+* name: identifier length
+  * description: The number of meaningful characters in the base identifier 
+  * type: integer
+  * range: 0,4096
+  * default:
+* name: case mapping
+  * description: Indicates how case in the source identifier should be handled when mapping to a path
+  * type: enumerated
+  * range: ToUpper,ToLower,Literal
+  * default:
+* name: invert mapping
+  * description: Indicate if mapping should begin at the rightmost (least significant) character of the stripped identifier
+  * type: enumerated
+  * range: FALSE,TRUE
+  * default: FALSE
+* name: tuple size
+  * description: Indicates the size of the chunks (in characters) that the identifier is split into during mapping
+  * type: integer
+  * range: 1,32
+  * default: 2
+* name: number of tuples
+  * description: Indicates how many chunks are used for path generation
+  * type: integer
+  * range: 0,32
+  * default:
+* short object root 
+  * description: Indicates how the OCFL object root directory name should be generated from the identifier
+  * type: enumerated
+  * range: FALSE,TRUE
+  * default: FALSE
+
+## Detailed explanation
+
+The approach described here is a generalisation of the [PairTree](https://tools.ietf.org/html/draft-kunze-pairtree-01) algorithm designed to be more flexible as file storage technologies have developed. Conventional filesystems are generally better able to handle large numbers of files in a directory and object stores tend to favour much flatter storage hierarchies. In short, the approach is to derive a unique file path for an OCFL object from its unique identifier in a programmatic and repeatable manner. 
+
+### identifier length
+
+This extension assumes that object unique identifiers are all the same length and that all the characters of the identifer are reasonably well distributed (e.g. for a hexadecimal-based identifier, each character can be any value from 0-f). This may mean that it is prudent to adjust the format of the identifier before it can be safely used to generate a path. For example, a [UUID](https://tools.ietf.org/html/rfc4122) is typically written in the form "uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6" where the "uuid:" premable and the hyphens are clearly non-unique elements. The "stripped" version used for path generation would be "f81d4fae7dec11d0a76500a0c91e6bf6". The **identifier length** parameter indicates the length of the stripped version of the identifier, in this example, 32.   
```

Comment:
I also don't like the assumption that ids are we distributed. Yes, that is ideal, but in the real world it is certainly not always the case. I think there needs to be an option for transforming object ids into a format that is suitable for storage. In fact, this is something that is already being assumed here when you talk about stripping off an identifier prefix. If my OCLF object's id is `uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6` where is it defined that I'm actually supposed to use `f81d4fae-7dec-11d0-a765-00a0c91e6bf6` when I'm trying to find the object?

I can understand not wanting to overly complicate this spec too much, but these are important details that need to be defined. Perhaps, id transformations should be handled in a separate spec? So, your OCFL storage layout is a combination of id translation and directory mapping?

Reply:
Yes, definitely a separate extension, or two! One that defines ID format, an optional one for translation to a well distributed form, and one that maps to a storage path.

Reply:
Created Issue #5

Reply:
Great! It's sounding like there could be multiple extensions involved in defining a repository layout. How do you specify multiple extensions in `ocfl_layout.json`?